training/controllers/mail.py

```diff
@@ -0,0 +1,134 @@
+from flask import Blueprint, request, render_template, url_for, redirect
+from flask_mail import Message
+from itsdangerous import URLSafeTimedSerializer, SignatureExpired
+import os
+
+from training.controllers.forms import new_password_form, recover_form, send_email_form
+from training.extensions import mail, bcrypt
+from training.models.users import User
+from training.extensions import db
+
+""" 
+    Problema original, solucionado. Sin embargo, estaria bueno discutir la solucion. Se relaciona con las background tasks para enviar mails por redis, ya que como 
+    me habias dicho Joaco, tenia que buscar una amanera de pasarle el app_context a redis. 
+
+    Mi problema original era el siguiente:
+    Este modulo es llamado cuando se esta a la mitad e create_app.
+    Recordemos que app, es creado como variable global al final de correrse create_app y aqui estamos en la mitad del flujo.
+    Por lo que cuando importamos "from training.background_tasks.tasks import send_email_async" este importa app desde app.py, pero aun no esta creada ya que
+    nos encontrabamos a mitad del flujo de create_app, y recordemos que app se crea cuando termina create_app. Ergo, ERROR
+    
+    Solucion:
+    Crear la app en 2 pasos. Se hace app=create_app(), y en create_app se inicializa, importa y crea todo salvo este modulo.
+    Luego, se llama a una funcion modifications(), que importa solo este modulo y se agrega el blueprint a la app ya existente.
+    
+    Entonces al importarse este modulo, se comienza a correr, y en determinado momento se importa app desde aca. 
+    Pero ya existe una app, por lo que no hay problema en ir a buscarla a app.py. Y se le puede dar el app_context a redis.
+"""
+
+
+s = URLSafeTimedSerializer(os.environ.get('SECRET_KEY'))
+
+bp = Blueprint('name3', __name__)
+
+#from training.background_tasks.tasks import send_email_async, send_email_function
+from training.app import app
+
+
+# Test mail
+@bp.route('/mail')
+def mail_test():
+    email = 'donnetta7@adriveriep.com'
+    msg = Message('Confirm Email', sender='anthony@prettyprinted.com', recipients=[email])
+    msg.body = "Bienvenido"
+    mail.send(msg)
+    return "0", 200
+
+
+# Validation process (DOES NOT SEND ANY EMAIL)
+@bp.route('/mail/validation/<string:username>')
+def email_verification(username):
+    our_user = User.query.filter_by(id=username).first()
+
+    if our_user is None:
+        return "User not exist", 450
+    else:
+        if our_user.mail_validation:
+            return "The email {} was already validated.".format(our_user.username), 205
+        else:
+            our_user.mail_validation = True
+            db.session.commit()
+            return "Email from user {} was successfully validated.".format(our_user.username), 200
+
+
+# Recover password process
+def send_email_function(msg):
+    with app.app_context():
+        # mail.suppress = True
+        mail.send(msg)
```

Comment:
Please put tasks in a standalone module, like background_tasks

Reply:
I was having issues with that, because if I did it into a separate module, I was not able to pass properly the app instance. However, i deeply researched what happen when i give a new thread a task who needs an app context. I will try to fix it in the next branch.